### PR TITLE
fix(wfctl): R-A4 align rule consults top-level secrets.generate + entries (workflow#541)

### PR DESCRIPTION
**Pull request by intel352**
## Summary

`wfctl infra align --strict` previously fired a false-positive `R-A4 FAIL` for env-var tokens that resolve via the **top-level** `secrets.generate` / `secrets.entries` blocks. `buildAlignContext` only populated `ctx.secretKeys` from the module-form (`secrets.generate` / `secrets.requires` modules), so any config using the canonical top-level `secrets:` block was unfairly flagged.

This PR fixes the original W-541 align bug. During Copilot review, three additional incomplete-merge surfaces in `config.processImports` were surfaced and addressed in follow-on commits to keep the import-merge logic internally consistent. Two further direct-parsing-caller paths (`parseSecretsConfig`, `wfctl secrets`) were scope-cut to follow-up issues for separate PRs.

Closes workflow#541. Unblocks core-dump PR #190 (which uses inline top-level `secrets.generate` to declare `STAGING_PG_PASSWORD`; verified against the actual config — no `imports:` involved).

## Scope (delta from original spec)

The original plan limited this PR to `buildAlignContext` + 2 align tests. Copilot review surfaced that `config.processImports` was missing several merge surfaces that the new align behavior would expose. Rather than ship a half-fix, the import-merge work was completed in the same PR. Honest scope:

### 1. Original W-541 fix — R-A4 align rule (commit `abc2db8c`)
- `cmd/wfctl/infra_align_rules.go`: extend `buildAlignContext`'s top-level-secrets branch to also seed `ctx.secretKeys` from `cfg.Secrets.Generate[].Key` and `cfg.Secrets.Entries[].Name`.
- Tests: `TestInfraAlign_RA4_TopLevelSecretsGenerate_DoesNotFire`, `TestInfraAlign_RA4_TopLevelSecretsEntries_DoesNotFire`.

### 2. processImports — Secrets merge (commit `8632612f`)
- `config/config.go`: `processImports` now merges `WorkflowConfig.Secrets`. Generate (dedupe by Key), Entries (dedupe by Name), scalars parent-wins.
- Tests: `TestLoadFromFile_ImportSecretsMerge`, `TestLoadFromFile_ImportSecretsOnlyInImport`, `TestInfraAlign_RA4_TopLevelSecrets_FromImport_DoesNotFire`.

### 3. processImports — SecretStores + Secrets.Config per-key merge (commit `dd81421d`)
- `config/config.go`: merge `WorkflowConfig.SecretStores` (per-store dedupe by name); change `Secrets.Config` from all-or-nothing to per-key merge to support shared-defaults + local-override pattern.
- Tests: `TestProcessImports_MergesSecretStoresFromImport`, `TestProcessImports_SecretsConfigMergesPerKey_LocalOverride`.

### 4. processImports — Environments merge + canonical provider IDs (commit `30bc8bc8`)
- `config/config.go`: merge `WorkflowConfig.Environments` per-env (parent wins). `ResolveSecretStore` consults `Environments[env].SecretsStoreOverride`; without merge, imported overrides drop and secrets route to wrong backend.
- Test fixtures: use canonical `aws-secrets-manager` and `gcp-secret-manager` (per `docs/dsl-reference.md` and `cmd/wfctl/wizard.go`).
- Tests: `TestProcessImports_MergesEnvironmentsFromImport`.

A complete `WorkflowConfig` field audit is documented in commit `30bc8bc8`'s body. Remaining unmerged top-level fields (Requires, Infrastructure, Engine, CI, Infra, Services, Mesh, Networking, Security) are intentionally out-of-scope follow-ups — none are referenced by the alignment-rule code path that motivated this PR.

## Follow-up issues

The merge work in this PR only benefits callers that go through `config.LoadFromFile`. Two direct-parsing-caller paths still bypass `processImports` and were scope-cut to keep this PR focused:

- workflow#551 — `parseSecretsConfig` (`cmd/wfctl/infra_bootstrap.go`, `cmd/wfctl/infra.go`) raw-unmarshals only the current file. After this PR, `wfctl infra align` correctly resolves imported `secrets.generate` keys, but `wfctl infra bootstrap` will silently skip generation of those secrets.
- workflow#552 — `wfctl secrets` family (`loadSecretsConfig`, `runSecretsSetup`) parses YAML directly. Configs that move `secrets.entries` / `secretStores` / per-env overrides into shared imports will look empty or route to the wrong store in `secrets get` / `validate` / `export` / `setup`.

Both are mechanical refactors (call `config.LoadFromFile` instead of `yaml.Unmarshal`) but have a wide enough blast radius to deserve their own PR + regression test coverage.

## Test plan

- [x] `GOWORK=off go test -race -count=1 ./config/ ./cmd/wfctl/` — green
- [x] Each new test verified to FAIL without its corresponding fix (proves bug + locks behavior)
- [x] Existing R-A4 module-form test (`TestInfraAlign_RA4_SecretsGenerate_DoesNotFire`) and all import precedence tests still pass
- [x] No `go.mod` / `go.sum` changes (the `Go Mod Tidy` failure on this PR is pre-existing AWS SDK drift on `main`, unrelated to this change)

## Design reference

`docs/plans/2026-05-05-iac-deferred-cleanup-design.md` (PR 1, W-541). v0.21.1 will be cut after this merges.

## Rollback

Revert the squash commit on `main`. Stopgap fallback is the env-var export pattern in core-dump's `deploy.yml` align step (`STAGING_PG_PASSWORD`, `STAGING_VPC_UUID`).

### CODE REVIEW

File: cmd/wfctl/infra_align_rules.go [hunk 1]
```diff
@@ -46,6 +46,12 @@ func buildAlignContext(cfgFile string) (*alignContext, error) {
 	}
 	if cfg.Secrets != nil {
 		ctx.secretGens = cfg.Secrets.Generate
+		for _, gen := range cfg.Secrets.Generate {
+			ctx.secretKeys[gen.Key] = struct{}{}
+		}
+		for _, entry := range cfg.Secrets.Entries {
+			ctx.secretKeys[entry.Name] = struct{}{}
+		}
```
[the remark this answers is not in the document]
Comment (intel352) [reply cmd/wfctl/infra_align_rules.go:54]: Real bug — confirmed and fixed in 8632612f. `processImports` previously merged modules / pipelines / triggers / platform / plugins / sidecars but not `WorkflowConfig.Secrets`, so an imported `secrets:` block was silently dropped. Now merges Generate (dedupe by Key) and Entries (dedupe by Name), with main-wins semantics on scalar/map fields (DefaultStore, Provider, Config, Rotation). New regression test `TestInfraAlign_RA4_TopLevelSecrets_FromImport_DoesNotFire` (cmd/wfctl) verified to FAIL without the merge — proves the bug. Plus two config-level tests `TestLoadFromFile_ImportSecretsMerge` and `TestLoadFromFile_ImportSecretsOnlyInImport` pin the merge semantics and the cfg.Secrets-nil-without-fix case.

File: cmd/wfctl/infra_align_test.go [hunk 2]
```diff
@@ -447,6 +447,63 @@ modules:
 	}
 }
 
+func TestInfraAlign_RA4_TopLevelSecretsGenerate_DoesNotFire(t *testing.T) {
+	os.Unsetenv("STAGING_PG_PASSWORD")
+	yaml := `
+appName: test
+secrets:
+  generate:
+    - key: STAGING_PG_PASSWORD
```
[the remark this answers is not in the document]
Comment (intel352) [reply cmd/wfctl/infra_align_test.go:456]: Addressed in 8632612f. New test `TestInfraAlign_RA4_TopLevelSecrets_FromImport_DoesNotFire` writes `main.yaml` + `shared.yaml` to t.TempDir, where `shared.yaml` declares both `secrets.generate` and `secrets.entries`, and `main.yaml` references the keys via `${...}` env vars. Verified to FAIL without the processImports merge fix (output: `unexpected R-A4 finding for imported top-level secrets: [{R-A4 FAIL api unresolved env var: ${STAGING_PG_PASSWORD}} {R-A4 FAIL api unresolved env var: ${STAGING_API_TOKEN}}]`). Now passes.

File: config/config.go [hunk 3]
```diff
@@ -361,6 +361,52 @@ func (cfg *WorkflowConfig) processImports(seen map[string]bool) error {
...
+			}
+			// Scalar/map fields: parent wins; only adopt if unset on parent.
+			if cfg.Secrets.DefaultStore == "" {
+				cfg.Secrets.DefaultStore = impCfg.Secrets.DefaultStore
+			}
+			if cfg.Secrets.Provider == "" {
+				cfg.Secrets.Provider = impCfg.Secrets.Provider
+			}
+			if cfg.Secrets.Config == nil {
+				cfg.Secrets.Config = impCfg.Secrets.Config
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go]: Real bug — fixed in dd81421d. processImports now merges WorkflowConfig.SecretStores with the same parent-wins dedupe-by-name semantics as Workflows / Triggers / Pipelines / Platform / Plugins.External. New test `TestProcessImports_MergesSecretStoresFromImport` covers (a) an imported store name surviving the merge and (b) parent-wins on a same-named store override. Verified to FAIL without the fix (`expected SecretStores[vault] from import`).

[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go]: Real bug — fixed in dd81421d. `Secrets.Config` now uses per-key merge instead of all-or-nothing. Imported keys absent from main survive; main wins on key conflicts. New test `TestProcessImports_SecretsConfigMergesPerKey_LocalOverride` mirrors your shared-defaults / local-override scenario: import provides {repo, token_env, api_url}, main only overrides {token_env}; after the merge cfg.Secrets.Config has all three keys with token_env=MAIN_TOKEN and repo/api_url surviving from the import. Verified to FAIL without the fix (`expected repo=GoCodeAlone/workflow (imported default), got <nil>`).

File: config/config_import_test.go [hunk 4]
```diff
@@ -641,3 +641,148 @@ modules:
...
+			t.Errorf("expected Entries name %q present after merge", n)
+		}
+	}
+	if shared, ok := entryByName["SHARED_ENTRY"]; ok && shared.Store != "main-store" {
+		t.Errorf("expected main-wins for SHARED_ENTRY store, got %q", shared.Store)
+	}
+}
+
+// TestLoadFromFile_ImportSecretsOnlyInImport covers the case the original
+// PR-1 fix missed: top-level secrets:`appears only in an imported file, not
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config_import_test.go]: Fixed in dd81421d — comment now reads `top-level \`secrets:\` appears only in an imported file` (with the missing space restored).

File: config/config.go [hunk 5]
```diff
@@ -361,6 +361,79 @@ func (cfg *WorkflowConfig) processImports(seen map[string]bool) error {
...
+				}
+			}
+		}
+
+		// Merge top-level secrets. Generate (dedupe by Key) and Entries
+		// (dedupe by Name) are appended. Scalar fields follow parent-wins.
+		// `Config` is a map[string]any: per-key merge so an imported "shared
+		// defaults" config can survive a partial main-file override (e.g.
+		// import provides {repo, token_env}; main only sets {token_env}).
+		if impCfg.Secrets != nil {
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go:404]: Real bug — fixed in 30bc8bc8. processImports now merges WorkflowConfig.Environments per-env (parent-wins on conflicts; imported envs absent from main are preserved). New test `TestProcessImports_MergesEnvironmentsFromImport` covers (a) imported env surviving (`staging.SecretsStoreOverride=vault`), (b) parent-wins on conflicting env (`production` main override beats import), (c) main-only env preserved. Verified to FAIL without the merge.

Also did a complete WorkflowConfig field audit (see commit body) so this is the last cascade round for #541. Remaining unmerged top-level fields (Requires, Infrastructure, Engine, CI, Infra, Services, Mesh, Networking, Security) are intentionally out-of-scope follow-ups — each needs its own merge-semantics design + regression tests, and none are referenced by the alignment-rule code path that motivated this PR.

File: config/config_import_test.go [hunk 6]
```diff
@@ -641,3 +641,261 @@ modules:
...
+	dir := t.TempDir()
+
+	importedYAML := `
+secretStores:
+  vault:
+    provider: vault
+    config:
+      address: https://vault.example.com
+  shared-store:
+    provider: aws-secretsmanager
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config_import_test.go]: Fixed in 30bc8bc8 — fixture now uses the canonical `aws-secrets-manager` (verified against docs/dsl-reference.md:947 and cmd/wfctl/wizard.go:144).

File: config/config_import_test.go [hunk 7]
```diff
@@ -641,3 +641,261 @@ modules:
...
+		t.Fatal(err)
+	}
+
+	mainYAML := `
+imports:
+  - imported.yaml
+
+secretStores:
+  shared-store:
+    provider: gcp-secretmanager
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config_import_test.go]: Fixed in 30bc8bc8 — fixture now uses the canonical `gcp-secret-manager` (verified against docs/dsl-reference.md:947 and cmd/wfctl/wizard.go:145). The assertion that checks parent-wins was also updated to match.

File: config/config.go [hunk 8]
```diff
@@ -361,6 +361,97 @@ func (cfg *WorkflowConfig) processImports(seen map[string]bool) error {
...
+					cfg.Environments[k] = v
+				}
+			}
+		}
+
+		// Merge top-level secrets. Generate (dedupe by Key) and Entries
+		// (dedupe by Name) are appended. Scalar fields follow parent-wins.
+		// `Config` is a map[string]any: per-key merge so an imported "shared
+		// defaults" config can survive a partial main-file override (e.g.
+		// import provides {repo, token_env}; main only sets {token_env}).
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go:403]: Scope-cut to follow-up issue workflow#551. Out of scope for W-541 which is core-dump unblock; the direct-parsing-caller refactor is a substantial enough change to warrant its own PR with proper review focus.

Verified core-dump's infra.yaml does not use `imports:` (its `secrets:` block is inline at the top level), so this PR's existing fix is sufficient to unblock C-1 PR #190. Filed #551 to track the parseSecretsConfig refactor for callers that go through bootstrap/plan paths.

File: config/config.go [hunk 9]
```diff
@@ -361,6 +361,97 @@ func (cfg *WorkflowConfig) processImports(seen map[string]bool) error {
...
+			}
+			for _, e := range impCfg.Secrets.Entries {
+				if _, exists := existingEntries[e.Name]; exists {
+					continue
+				}
+				cfg.Secrets.Entries = append(cfg.Secrets.Entries, e)
+				existingEntries[e.Name] = struct{}{}
+			}
+		}
 	}
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go:456]: Scope-cut to follow-up issue workflow#552. Out of scope for W-541 which is core-dump unblock; the direct-parsing-caller refactor is a substantial enough change to warrant its own PR with proper review focus.

Filed #552 to track the `wfctl secrets` family migration to `config.LoadFromFile` so the import-merge work in this PR is honored across get / validate / export / setup paths.

File: config/config.go [hunk 10]
```diff
@@ -361,6 +361,97 @@ func (cfg *WorkflowConfig) processImports(seen map[string]bool) error {
 			cfg.Sidecars = append(cfg.Sidecars, sc)
 			existingSidecars[sc.Name] = struct{}{}
 		}
+
+		// Merge SecretStores — per-store dedupe by name (parent wins).
+		// SecretsConfig.DefaultStore + SecretEntry.Store reference these by
+		// name via ResolveSecretStore / getProviderForStore, so the import
+		// merge must include the store map; otherwise an imported store name
+		// is later treated as a raw provider and provider construction fails.
```
[the remark this answers is not in the document]
Comment (intel352) [reply config/config.go:369]: Acknowledged — updating PR description to reflect actual scope (R-A4 align fix + processImports merge completeness for Secrets / SecretStores / Secrets.Config / Environments) and link the two follow-up issues for the direct-parsing-caller refactors (#551, #552).
